flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestIcebergSourceSplitReader.java

```diff
@@ -0,0 +1,167 @@
+
+  @Test
+  public void testPause() throws Exception {
+    IcebergSourceSplitReader<RowData> reader = reader();
+
+    List<List<Record>> records =
+        ImmutableList.of(
+            RandomGenericData.generate(SCHEMA, 2, 0L), RandomGenericData.generate(SCHEMA, 3, 1L));
+
+    IcebergSourceSplit split =
+        ReaderUtil.createSplit(records, TEMPORARY_FOLDER, FileFormat.PARQUET, APPENDER_FACTORY);
+
+    // Add the new split to the reader
+    reader.handleSplitsChanges(new SplitsAddition<>(ImmutableList.of(split)));
+
+    // Fetch the first batch, and check the result size
+    ArrayBatchRecords<RecordAndPosition<RowData>> result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(2);
+
+    // Pause the reading of the split, and check that the read is blocked when trying to fetch new
+    // records
+    reader.pauseOrResumeSplits(ImmutableList.of(split), ImmutableList.of());
+    assertThatThrownBy(
+            () ->
+                Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> reader.fetch() != null))
+        .isInstanceOf(ConditionTimeoutException.class)
+        .hasMessageContaining("was not fulfilled within");
+
+    // Unpause the reading of the split, and check the result size
+    reader.pauseOrResumeSplits(ImmutableList.of(), ImmutableList.of(split));
+    result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(3);
+  }
+
+  @Test
+  public void testWakeup() throws Exception {
+    IcebergSourceSplitReader<RowData> reader = reader();
+
+    List<List<Record>> records =
+        ImmutableList.of(
+            RandomGenericData.generate(SCHEMA, 2, 0L), RandomGenericData.generate(SCHEMA, 3, 1L));
+    IcebergSourceSplit split =
+        ReaderUtil.createSplit(records, TEMPORARY_FOLDER, FileFormat.PARQUET, APPENDER_FACTORY);
+
+    // Add the new split to the reader
+    reader.handleSplitsChanges(new SplitsAddition<>(ImmutableList.of(split)));
+
+    // Fetch the first batch, and check the result size
+    ArrayBatchRecords<RecordAndPosition<RowData>> result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(2);
+
+    // Pause the reading of the split, and check that the read is blocked when trying to fetch new
+    // records
+    reader.pauseOrResumeSplits(ImmutableList.of(split), ImmutableList.of());
+
+    assertThatThrownBy(
+            () ->
+                Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> reader.fetch() != null))
+        .isInstanceOf(ConditionTimeoutException.class)
+        .hasMessageContaining("was not fulfilled within");
+
+    // Run the new fetch and wakeUp concurrently. Expect empty result from the blocked fetch.
+    ExecutorService executorService =
+        MoreExecutors.getExitingExecutorService(
+            (ThreadPoolExecutor) Executors.newFixedThreadPool(1));
+
+    AtomicBoolean emptyResult = new AtomicBoolean(false);
```

Comment:
nit: `nonemptyResult` or otherwise `emptyResult` should be true initially.

Reply:
Renamed to `gotEmptyResult`